File: app/src/main/java/uk/co/sundroid/util/view/BoundedLinearLayout.java
package uk.co.sundroid.util.view;

import uk.co.sundroid.R;
import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.widget.LinearLayout;

public class BoundedLinearLayout extends LinearLayout {

	private final int mBoundedWidth;

	private final int mBoundedHeight;

	public BoundedLinearLayout(Context context) {
		super(context);
		mBoundedWidth = 0;
		mBoundedHeight = 0;
	}

	public BoundedLinearLayout(Context context, AttributeSet attrs) {
		super(context, attrs);
		TypedArray a = getContext().obtainStyledAttributes(attrs,
				R.styleable.BoundedLinearLayout);
		mBoundedWidth = a.getDimensionPixelSize(
				R.styleable.BoundedLinearLayout_bounded_width, 0);
		mBoundedHeight = a.getDimensionPixelSize(
				R.styleable.BoundedLinearLayout_bounded_height, 0);
		a.recycle();
	}

	@Override
	protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
		// Adjust width as necessary
		int measuredWidth = MeasureSpec.getSize(widthMeasureSpec);
		if (mBoundedWidth > 0 && mBoundedWidth < measuredWidth) {
			int measureMode = MeasureSpec.getMode(widthMeasureSpec);
			widthMeasureSpec = MeasureSpec.makeMeasureSpec(mBoundedWidth,
					measureMode);
		}
		// Adjust height as necessary
		int measuredHeight = MeasureSpec.getSize(heightMeasureSpec);
		if (mBoundedHeight > 0 && mBoundedHeight < measuredHeight) {
			int measureMode = MeasureSpec.getMode(heightMeasureSpec);
			heightMeasureSpec = MeasureSpec.makeMeasureSpec(mBoundedHeight,
					measureMode);
		}
		super.onMeasure(widthMeasureSpec, heightMeasureSpec);
	}
}

File: app/src/main/kotlin/uk/co/sundroid/util/view/BoundedLinearLayout.kt
package uk.co.sundroid.util.view

import uk.co.sundroid.R
import android.content.Context
import android.util.AttributeSet
import android.widget.LinearLayout

class BoundedLinearLayout(context: Context, attrs: AttributeSet) : LinearLayout(context, attrs) {

    private val boundedWidth: Int
    private val boundedHeight: Int

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        // Adjust width as necessary
        var widthSpec = widthMeasureSpec
        val measuredWidth = MeasureSpec.getSize(widthSpec)
        if (boundedWidth in 1..(measuredWidth - 1)) {
            val measureMode = MeasureSpec.getMode(widthSpec)
            widthSpec = MeasureSpec.makeMeasureSpec(boundedWidth, measureMode)
        }
        // Adjust height as necessary
        var heightSpec = heightMeasureSpec
        val measuredHeight = MeasureSpec.getSize(heightSpec)
        if (boundedHeight in 1..(measuredHeight - 1)) {
            val measureMode = MeasureSpec.getMode(heightSpec)
            heightSpec = MeasureSpec.makeMeasureSpec(boundedHeight, measureMode)
        }
        super.onMeasure(widthSpec, heightSpec)
    }

    init {
        val array = context.obtainStyledAttributes(attrs, R.styleable.BoundedLinearLayout)
        boundedWidth = array.getDimensionPixelSize(R.styleable.BoundedLinearLayout_bounded_width, 0)
        boundedHeight = array.getDimensionPixelSize(R.styleable.BoundedLinearLayout_bounded_height, 0)
        array.recycle()
    }
}
